spec/forms/keywords_form_spec.rb

```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe KeywordsForm, type: :form do
+  describe '#save' do
+    context 'given a valid file it' do
+      it 'accepts a valid CSV' do
+        form = save_file('valid.csv')
+        expect(form.errors).to be_empty
+      end
+    end
+
+    context 'given an invalid file, does not accept the file if it' do
+      it 'has no file being uploaded' do
+        form = save_file(nil)
+        expect(form.errors.full_messages.first).to eq I18n.t('csv.upload_fail_missing_file')
+      end
+
+      it 'has less than 1 word' do
+        form = save_file('0_words.csv')
+        expect(form.errors.full_messages.first).to eq I18n.t('csv.upload_fail_invalid_size')
+      end
+
+      it 'has more than 1000 words' do
+        form = save_file('1001_words.csv')
+        expect(form.errors.full_messages.first).to eq I18n.t('csv.upload_fail_invalid_size')
+      end
+
+      it 'does not have a CSV extension' do
+        form = save_file('invalid_file_ext.txt')
+        expect(form.errors.full_messages.first).to eq I18n.t('csv.upload_fail_invalid_extention')
+      end
+    end
```

Comment:
We can split this case out of the the `given an invalid file`, right?

```suggestion
    context 'given NO file' do
      it 'has no file being uploaded' do
        form = save_file(nil)

        expect(form.errors.full_messages.first).to eq I18n.t('csv.upload_fail_missing_file')
      end
    end

    context 'given an invalid file' do
      context 'given file is empty' do
        it 'adds an invalid size error' do
          form = save_file('0_words.csv')

          expect(form.errors.full_messages.first).to eq I18n.t('csv.upload_fail_invalid_size')
        end
      end

      context 'given file has more than 1000 words' do
        it 'adds an invalid size error' do
          form = save_file('1001_words.csv')
          expect(form.errors.full_messages.first).to eq I18n.t('csv.upload_fail_invalid_size')
        end
      end

      context 'given file does NOT have a CSV extension' do
        it 'adds an invalid extension error' do
          form = save_file('invalid_file_ext.txt')
          expect(form.errors.full_messages.first).to eq I18n.t('csv.upload_fail_invalid_extention')
        end
      end
    end
```

Reply:
Updated in https://github.com/aetheryna/google-scraper-ruby/pull/34/commits/25b9f583ac59d64e0e10127480a5f768837eeeaa 😎